src/environment.js

```diff
@@ -959,16 +966,28 @@ export class GlobalEnvironmentRecord extends EnvironmentRecord {
 }
 
 // ECMA262 8.1
+let uid = 0;
 export class LexicalEnvironment {
   constructor(realm: Realm) {
     invariant(realm, "expected realm");
     this.realm = realm;
+    this.destroyed = false;
+    this._sno = uid++;
   }
 
+  _sno: number;
+  destroyed: boolean;
   environmentRecord: EnvironmentRecord;
   parent: null | LexicalEnvironment;
   realm: Realm;
 
+  destroy() {
+    this.destroyed = true;
+    if (this.environmentRecord instanceof DeclarativeEnvironmentRecord) {
+      this.environmentRecord.frozen = true;
```

Comment:
This needs some explanation. Destruction and freezing do not necessarily go together.